packages/wallet/src/TxBuilder/types.ts

```diff
@@ -0,0 +1,126 @@
+import { Cardano } from '@cardano-sdk/core';
+import { CustomError } from 'ts-custom-error';
+import { DeepPartial } from '@cardano-sdk/util';
+import { InputSelectionError } from '@cardano-sdk/cip2';
+
+import { OutputValidation } from '../types';
+import { SignTransactionOptions } from '../KeyManagement';
+
+// Does not transform TokenMap into Partial<TokenMap> because it would also make it's methods and map types optional
+// and that forces users to do many checks before being able to use any of it's properties and methods
+export type PartialTxOut = DeepPartial<Cardano.TxOut, Cardano.TokenMap>;
+
+export enum TxOutputFailure {
+  MinimumCoin = 'Minimum coin not met',
+  TokenBundleSizeExceedsLimit = 'Token Bundle Exceeds Limit',
+  MissingRequiredFields = 'Mandatory fields address or coin are missing'
+}
+
+export class OutputValidationMissingRequiredError extends CustomError {
+  public constructor(public txOut: PartialTxOut) {
+    super(TxOutputFailure.MissingRequiredFields);
+  }
+}
+
+export class OutputValidationMinimumCoinError extends CustomError {
+  public constructor(public txOut: PartialTxOut, public outputValidation: OutputValidation) {
+    super(TxOutputFailure.MinimumCoin);
+  }
+}
+
+export class OutputValidationTokenBundleSizeError extends CustomError {
+  public constructor(public txOut: PartialTxOut, public outputValidation: OutputValidation) {
+    super(TxOutputFailure.TokenBundleSizeExceedsLimit);
+  }
+}
+
+export class IncompatibleWalletError extends CustomError {}
+
+export type TxOutValidationError =
+  | OutputValidationMissingRequiredError
+  | OutputValidationMinimumCoinError
+  | OutputValidationTokenBundleSizeError;
+export type TxBodyValidationError = TxOutValidationError | InputSelectionError;
+
+export type Valid<TValid> = TValid & {
+  isValid: true;
+};
+
+export interface Invalid<TError> {
+  isValid: false;
+  errors: TError[];
+}
+
+export interface ValidTxOutData {
+  readonly txOut: Cardano.TxOut;
+}
+
+export type ValidTxOut = Valid<ValidTxOutData>;
+export type InvalidTxOut = Invalid<TxOutValidationError>;
+export type MaybeValidTxOut = ValidTxOut | InvalidTxOut;
+
+export interface OutputBuilder {
```

Comment:
More doc comments on this type would be great

Reply:
I've added doc comments in the concrete implementations. Should I hoist them to the interfaces?

Reply:
Yes please, I think concrete implementations inherit docs from the interface. Seems to be working with vscode intellisense, although not sure if it gets propagated to the generated docs of concrete implementations. If it doesn't then we might want to have the doc comments duplicated.

Reply:
I have hoisted the docs to the interface here: [58a49fb](https://github.com/input-output-hk/cardano-js-sdk/pull/396/commits/58a49fb87bfb7d17aa07f9c26147f9ffcff70cd7).
The generated documentation propagates nicely to the concrete classes too.